src/quantuminspire/projectq/backend_qx.py

```diff
@@ -303,65 +283,105 @@ def _run(self):
         if self.qasm == "":
             return
 
-        if self._verbose:
-            print('_run (id {0})'.format((id(self), )))
-
         # finally: add measurement commands for all measured qubits if no measurements are given.
         # only measurements after all gate operations will perform properly
         if not self._measured_ids:
             self.__add_measure_all_qubits()
 
-        if self._verbose >= 2:
-            print('_run: self._allocated_qubits {0}'.format((self._allocated_qubits, )))
-        max_qubit_id = self._max_qubit_id
+        self._finalize_qasm()
+        self._execute_cqasm()
+        self._filter_result_by_measured_qubits()
+        self._register_random_measurement_outcome()
+        self._reset()
 
+    def _finalize_qasm(self):
+        """ Finalize qasm (add version and qubits line) """
         qasm = 'version 1.0\n# generated by Quantum Inspire {0} class\nqubits {1}\n\n'.format(
-            self.__class__, max_qubit_id+1)
+            self.__class__, self._number_of_qubits)
         qasm += self.qasm
 
-        try:
-            if self._verbose:
-                print('sending cqasm:')
-                print('------')
-                print(qasm)
-                print('------')
+        self._cqasm = qasm
+
+    def _execute_cqasm(self):
+        """ Execute self._cqasm, raise error if raw_text in result from API is not empty (which indicates an error)
```

Comment:
> To avoid google docstring nitpicking, I do believe you should describe errors raised with Raises: and then list the specific error classes raised and under what condition

Reply:
:cactus: (sorry, couldn't find the checkmark, and this was the first emoticon to show when typing colon followed by c)